### ISSUE TITLE
[Tool] Visual Studio Code Plugin, Language Server, and Grammar

### ISSUE THREAD

**Issue by achimmihca** [UltraStar-Deluxe/format#18]
### Suggestion

It would be nice to have a VisualStudio Code plugin for the UltraStar txt file format.

- The plugin could suggest available header fields defined by this format.
- The plugin could highlight errors in the format, e.g., overlapping notes, out of order notes
- The plugin could suggest improvements, e.g., adding golden notes

Support for a new language in Visual Studio Code (and many other IDEs) can be added via a [Language Server](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide).

### Grammar of the Format

- The Language Server could be implemented by using a parser generator (e.g. [ANTLR](https://www.antlr.org/)) and a grammar of the format.
    - See https://neuroning.com/post/implementing-code-completion-for-vscode-with-antlr/
    - See https://tomassetti.me/autocompletion-editor-antlr/
- The grammar would provide a formal specification of the format.
- Being able to create a parser from a grammar might improve the quality of the format.
    - Stuff that is hard to parse or formulate in the grammar would be found before implementing it in USDX and other tools. And such stuff is probably not a good idea in the first place.

### Use case

People editing the format by hand in text editor could use the VisualStudio Code plugin.

**Comment by Baklap4** [comment]
Would be a really nice addition, but i doubt people are opening the txt files within VSCode that much (at least not your average player)

> The plugin could suggest available header fields defined by this format.

Yes love this. Including some documentation about the field or note.

> The plugin could highlight errors in the format, e.g., overlapping notes, out of order notes

That'd be really nice! This helps spotting errors early on. Not sure if possible, but implement this checking also on txt-upload

> The plugin could suggest improvements, e.g., adding golden notes

Yes things get forgotten, so suggesting improvements is really nice.


Just a question would it be possible to store this in a library or something so creator software can also use this instead of just VSCode?

**Comment by achimmihca** [comment]
> would it be possible to store this in a library or something so creator software can also use this instead of just VSCode

Yes, by implementing the language server protocol it is editor agnostic (see [illustration](https://code.visualstudio.com/assets/api/language-extensions/language-server-extension-guide/lsp-languages-editors.png))

**Comment by codello** [comment]
#48 includes a first proposal for a grammar (in ABNF). There are still some open questions about some edge cases but it‘s a start.

I like the idea of having more tooling around the file format but I‘m not sure it would be in scope for this project. Maybe having a VS code plugin and a language server be separate projects would be a better fit?
